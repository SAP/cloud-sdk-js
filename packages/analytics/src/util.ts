/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import crypto from 'crypto';

/**
 * Generate a random salt.
 * @returns A random salt.
 * @hidden
 */
export function randomSalt(): string {
  // 1 byte will produce two bytes in hex string
  // 32 random bytes generate a hex string of the same bitlength as SHA-256
  return crypto.randomBytes(32).toString('hex');
}

/**
 * Create a hash for the given value.
 * @param value - The value to hash.
 * @returns The hashed value.
 * @hidden
 */
export function hash(value: string): string {
  return crypto
    .createHash('sha256')
    .update(value)
    .digest('hex');
}
