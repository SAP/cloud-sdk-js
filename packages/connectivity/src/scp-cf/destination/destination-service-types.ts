import type { CachingOptions } from '../cache';
import type { ProxyConfiguration } from '../connectivity-service-types';
import type { ResilienceOptions } from '../resilience-options';
import type { IsolationStrategy } from './destination-cache';

/**
 * A resolved destination containing information needed to execute requests, such as the system URL.
 *
 * You can create a destination as a local object when supplying all necessary information, or it could be retrieved from the destination service on SAP Business Technology Platform (via {@link DestinationFetchOptions}).
 * When creating a local object representing a destination, you need to supply at least the {@link url} and, if required by the target system, valid credentials with {@link username} and {@link password}.
 */
export interface Destination {
  /**
   * Name of the destination retrieved from SAP Business Technology Platform.
   */
  name?: string | null;

  /**
   * Base URL for calls to this destination.
   * The URL has to define the protocol, like `http://` or `https://`, and a host.
   * The path for requests against this destination will be appended to the path defined in the URL as a new path segment.
   */
  url: string;

  /**
   * Type of authentication to use.
   *
   * Defaults to `NoAuthentication`, unless {@link username} and {@link password} are provided, in which case the default is `BasicAuthentication`.
   */
  authentication?: AuthenticationType;

  /**
   * Proxy type to specify whether the target resides on-premise (not used).
   */
  proxyType?: DestinationProxyType;

  /**
   * Client to target in an SAP system, will be added as HTTP header `sap-client` if set.
   */
  sapClient?: string | undefined | null;

  /**
   * Username to use for basic authentication, optional if other means of authentication shall be used.
   */
  username?: string | null;

  /**
   * Password to use for basic authentication, optional if other means of authentication shall be used.
   */
  password?: string | null;

  /**
   * Authentication tokens returned from destination service on SAP Business Technology Platform.
   */
  authTokens?: DestinationAuthToken[] | null;

  /**
   * Flag indicating whether all certificates should be accepted when communicating with the destination. Should not be "true" in production.
   */
  isTrustingAllCertificates?: boolean;

  /**
   * ProxyConfiguration for on-premise connectivity and http(s) web proxies. Is present if proxyType of the destination equals "OnPremise" or environment variables [http_proxy] or [https_proxy] are set See {@link ProxyConfiguration}.
   */
  proxyConfiguration?: ProxyConfiguration;

  /**
   * Client Id used to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  clientId?: string;

  /**
   * Client Secret used to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  clientSecret?: string;

  /**
   * URL to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  tokenServiceUrl?: string;

  /**
   * User for basic authentication to OAuth server (if required).
   */
  tokenServiceUser?: string;

  /**
   * Password for tokenServiceUser (if required).
   */
  tokenServicePassword?: string;

  /**
   * The type of the destination, defaults to 'HTTP'. The SAP Cloud SDK only understands destinations of type 'HTTP' and 'MAIL'.
   */
  type?: 'HTTP' | 'LDAP' | 'MAIL' | 'RFC';

  /**
   * Further properties of the destination as defined in destination service on SAP Business Technology Platform, possibly empty.
   */
  originalProperties?: { [key: string]: any };

  /**
   * Flag indicating whether the destination is for test purpose. Should be "undefined" or "false" for non-mocked destinations.
   */
  isTestDestination?: boolean;

  /**
   * Location ID of the Cloud Connector to be used for connection to an On-Premise system. Optional. Corresponds to property "CloudConnectorLocationId" in the additional properties of a destination.
   */
  cloudConnectorLocationId?: string;

  /**
   * Array of certificates used for authentication type ClientCertificateAuthentication.
   */
  certificates?: DestinationCertificate[];

  /**
   * Trusted certificate for https requests used by the destination. Only relevant for self-signed certificates where trust needs to be enabled in the client.
   */
  trustStoreCertificate?: DestinationCertificate;

  /**
   * Name of the key store/certificate to be used for ClientCertificateAuthentication.
   */
  keyStoreName?: string;

  /**
   * Password of the key store/certificate to be used for ClientCertificateAuthentication.
   */
  keyStorePassword?: string;

  /**
   * System user to be used for OAuth2SAMLBearerAssertion authentication type.
   */
  systemUser?: string;

  /**
   * Additional headers to be used for calls against the destination, originally defined by `URL.headers.<header-name>`.
   * The keys of this object denote the names of the headers and the values their values.
   */
  headers?: Record<string, any>;

  /**
   * Additional query parameters to be used for calls against the destination, originally defined by `URL.queries.<query-parameter-name>`.
   * The keys of this object denote the names of the query parameters and the values their values.
   */
  queryParameters?: Record<string, any>;

  /**
   * If set to true the auth token provided to the request execution is forwarded to the destination target.
   */
  forwardAuthToken?: boolean;

  /**
   * Base64-encoded JSON web key set, containing the signing keys which are used to validate the JWT provided in the X-User-Token header.
   * This field is used to authenticate the destination using a JWT without JKU.
   */
  jwks?: string;

  /**
   * URI of the JSON web key set, containing the signing keys which are used to validate the JWT provided in the X-User-Token header.
   * This field is used to authenticate the destination using a JWT without JKU.
   */
  jwksUri?: string;
}

/**
 * Represents authentication token returned from destination service.
 */
export interface DestinationAuthToken {
  /**
   * Type of the token, e.g., Bearer.
   */
  type: string;
  /**
   * Value of the token.
   */
  value: string;
  /**
   * The number of seconds until the access token expires.
   */
  expiresIn: string;
  /**
   * Potential error of token retrieval in the destination service.
   */
  error: string | null;
  /**
   * An object containing ready-to-use key-value pairs like `Authorization` : `Bearer U29tZVRva2VuVmFsdWU=`.
   */
  http_header: {
    key: string;
    value: string;
  };
}

/**
 * Represents the proxy type on a destination.
 */
export type DestinationProxyType =
  | 'OnPremise'
  | 'Internet'
  | 'PrivateLink'
  | null;

/**
 * Represents a certificate attached to a destination.
 */
export interface DestinationCertificate {
  /**
   * Name of the certificate file.
   */
  name: string;

  /**
   * Content of the certificate as base64 encoded binary.
   */
  content: string;

  /**
   * Type of the certificate.
   */
  type: string;
}

/**
 * Options to use while fetching destinations. Encompasses both {@link CachingOptions} and {@link ResilienceOptions} interfaces.
 */
export type DestinationRetrievalOptions = CachingOptions &
  ResilienceOptions & {
    /**
     * The isolation strategy used for caching destinations. For the available options, see {@link IsolationStrategy}.
     * By default, IsolationStrategy.Tenant_User is set.
     */
    isolationStrategy?: IsolationStrategy;
  };

/**
 * Typeguard to find if object is a Destination.
 * @param destination - Destination to be checked
 * @returns boolean
 * @internal
 */
export function isDestination(destination: any): destination is Destination {
  return destination.url !== undefined;
}

/**
 * Represents the authentication type of a destination.
 */
export type AuthenticationType =
  | 'PrincipalPropagation'
  | 'NoAuthentication'
  | 'BasicAuthentication'
  | 'SAMLAssertion'
  | 'OAuth2SAMLBearerAssertion'
  | 'OAuth2ClientCredentials'
  | 'OAuth2UserTokenExchange'
  | 'ClientCertificateAuthentication'
  | 'OAuth2JWTBearer'
  | 'OAuth2Password';

/**
 * The destinations endpoint distinguished between destinations maintained on service level (instance) and account level (subaccount).
 * This enum is used as a switch in the {@link fetchInstanceDestinations}, {@link fetchSubaccountDestinations}  and {@link destinationServiceCache}
 * @internal
 */
export enum DestinationType {
  Instance = 'instance',
  Subaccount = 'subaccount'
}
