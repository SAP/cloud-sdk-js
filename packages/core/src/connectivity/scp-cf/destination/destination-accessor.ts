import { VerifyJwtOptions } from '../jwt';
import { sanitizeDestination } from './destination';
import {
  Destination,
  DestinationNameAndJwt,
  DestinationRetrievalOptions,
  isDestinationNameAndJwt
} from './destination-service-types';
import { searchEnvVariablesForDestination } from './destination-from-env';
import { searchServiceBindingForDestination } from './destination-from-vcap';
import { getDestinationFromDestinationService } from './destination-from-service';
import { DestinationAccessorOptions } from './destination-accessor-types';

/**
 * Returns the parameter if it is a destination, calls [[getDestination]] otherwise (which will try to fetch the destination
 * from the Cloud Foundry destination service).
 *
 * Fetching a destination requires:
 * - a binding to exactly one XSUAA service instance with service plan "application"
 * - a binding to a destination service instance
 *
 * If either of the prerequisites is not met or one of the services returns an error, this function will either throw an error or return a promise that rejects.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @param options - Caching options by fetching destination.
 * @returns A promise resolving to the requested destination on success.
 */
export async function useOrFetchDestination(
  destination: Destination | DestinationNameAndJwt,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return isDestinationNameAndJwt(destination)
    ? getDestination(destination.destinationName, {
        userJwt: destination.jwt,
        ...options
      })
    : sanitizeDestination(destination);
}

export type DestinationOptions = DestinationAccessorOptions &
  DestinationRetrievalOptions &
  VerifyJwtOptions;

/**
 * @deprecated Since v1.0.1. Use [[getDestination]] instead.
 *
 * Retrieves a destination with the given name from the Cloud Foundry destination service.
 * Returns `null`, if no destination can be found.
 * Requires the following service bindings: destination, XSUAA
 * By default, selects subscriber over provider and instance over subaccount destinations.
 *
 * If the destinations are read from the environment, the jwt will be ignored.
 * @param name - The name of the destination to be retrieved.
 * @param options - The options of the fetching query of the destination that include the JWT of the current request and the strategy for selecting a destination.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestinationOptions(
  name: string,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return getDestination(name, options);
}

/**
 * Builds a destination from one of three sources (in the given order):
 * - from the environment variable "destinations"
 * - from service bindings
 * - from the destination service
 *
 * If you want to get a destination only from a specific source, use the corresponding function directly
 *  (`getDestinationFromEnvByName`, `destinationForServiceBinding`, `getDestinationFromDestinationService`).
 * @param name - The name of the destination to be retrieved.
 * @param options - Configuration for how to retrieve destinations from the destination service.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestination(
  name: string,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return (
    searchEnvVariablesForDestination(name) ||
    searchServiceBindingForDestination(name) ||
    getDestinationFromDestinationService(name, options)
  );
}
