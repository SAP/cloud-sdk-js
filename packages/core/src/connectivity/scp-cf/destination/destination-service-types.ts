import { CachingOptions, IsolationStrategy } from '../cache';
import { ProxyConfiguration } from '../connectivity-service-types';
import { ResilienceOptions } from '../resilience-options';

/**
 * A resolved destination containing information needed to execute requests, such as the system URL.
 *
 * You can create a destination as a local object when supplying all necessary information, or it could be retrieved from the destination service on SAP Cloud Platform (via [[DestinationNameAndJwt]]).
 * When creating a local object representing a destination, you need to supply at least the [[url]] and, if required by the target system, valid credentials with [[username]] and [[password]].
 */
export interface Destination {
  /**
   * Name of the destination retrieved from SAP Cloud Platform.
   */
  name?: string | null;

  /**
   * Base URL for calls to this destination.
   * The URL has to define the protocol, like `http://` or `https://`, and a host.
   * The path for requests against this destination will be appended to the path defined in the URL as a new path segment.
   */
  url: string;

  /**
   * Type of authentication to use.
   *
   * Defaults to `NoAuthentication`, unless [[username]] and [[password]] are provided, in which case the default is `BasicAuthentication`.
   */
  authentication?: AuthenticationType;

  /**
   * Proxy type to specify whether the target resides on-premise (not used).
   */
  proxyType?: DestinationProxyType;

  /**
   * @deprecated Since v1.0.1.
   * Origin of the destination in a multi-tenant setup on SAP Cloud Platform (either from provider or subscriber account), optional.
   */
  origin?: DestinationOrigin;

  /**
   * Client to target in an SAP system, will be added as HTTP header `sap-client` if set.
   */
  sapClient?: string | undefined | null;

  /**
   * Username to use for basic authentication, optional if other means of authentication shall be used.
   */
  username?: string | null;

  /**
   * Password to use for basic authentication, optional if other means of authentication shall be used.
   */
  password?: string | null;

  /**
   * Authentication tokens returned from destination service on SAP Cloud Platform.
   */
  authTokens?: DestinationAuthToken[] | null;

  /**
   * Flag indicating whether all certificates should be accepted when communicating with the destination. Should not be "true" in production.
   */
  isTrustingAllCertificates?: boolean;

  /**
   * ProxyConfiguration for on-premise connectivity and http(s) web proxies. Is present if proxyType of the destination equals "OnPremise" or environment variables [http_proxy] or [https_proxy] are set See [[ProxyConfiguration]].
   */
  proxyConfiguration?: ProxyConfiguration;

  /**
   * Client Id used to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  clientId?: string;

  /**
   * Client Secret used to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  clientSecret?: string;

  /**
   * URL to retrieve access token for "OAuth2ClientCredentials", "OAuth2UserTokenExchange" and "OAuth2JWTBearer" authentication.
   */
  tokenServiceUrl?: string;

  /**
   * User for basic authentication to OAuth server (if required).
   */
  tokenServiceUser?: string;

  /**
   * Password for tokenServiceUser (if required).
   */
  tokenServicePassword?: string;

  /**
   * The type of the destination, defaults to 'HTTP'. The SAP Cloud SDK only understands destinations of type 'HTTP'.
   */
  type?: 'HTTP' | 'LDAP' | 'MAIL' | 'RFC';

  /**
   * Further properties of the destination as defined in destination service on SAP Cloud Platform, possibly empty.
   */
  originalProperties?: { [key: string]: any };

  /**
   * Flag indicating whether the destination is for test purpose. Should be "undefined" or "false" for non-mocked destinations.
   */
  isTestDestination?: boolean;

  /**
   * Location ID of the Cloud Connector to be used for connection to an On-Premise system. Optional. Corresponds to property "CloudConnectorLocationId" in the additional properties of a destination.
   */
  cloudConnectorLocationId?: string;

  /**
   * Array of certificates used for authentication type ClientCertificateAuthentication.
   */
  certificates?: DestinationCertificate[];

  /**
   * Name of the key store/certificate to be used for ClientCertificateAuthentication.
   */
  keyStoreName?: string;

  /**
   * Password of the key store/certificate to be used for ClientCertificateAuthentication.
   */
  keyStorePassword?: string;

  /**
   * System user to be used for OAuth2SAMLBearerAssertion authentication type.
   */
  systemUser?: string;

  /**
   * Additional headers to be used for calls against the destination, originally defined by `URL.headers.<header-name>`.
   * The keys of this object denote the names of the headers and the values their values.
   */
  headers?: Record<string, any>;

  /**
   * Additional query parameters to be used for calls against the destination, originally defined by `URL.queries.<query-parameter-name>`.
   * The keys of this object denote the names of the query parameters and the values their values.
   */
  queryParameters?: Record<string, any>;

  /**
   * If set to true the auth token provided to the request execution is forwarded to the destination target.
   */
  forwardAuthToken?: boolean;
}

/**
 * Represents authentication token returned from destination service.
 */
export interface DestinationAuthToken {
  type: string;
  value: string;
  expiresIn: string;
  error: string | null;
  http_header: {
    key: string;
    value: string;
  };
}

export type DestinationProxyType = 'OnPremise' | 'Internet' | null;

/**
 * Represents a certificate attached to a destination.
 */
export interface DestinationCertificate {
  /**
   * Name of the certificate file.
   */
  name: string;

  /**
   * Content of the certificate as base64 encoded binary.
   */
  content: string;

  /**
   * Type of the certificate.
   */
  type: string;
}

/**
 * @deprecated Since v1.0.1.
 *
 * Represents the origin of a destination in a multi-tenant setup on SAP Cloud Platform.
 *
 * In a multi-tenant application on SAP Cloud Platform, destinations can be defined both on provider account level ("PaaS tenant") as well as on the level of each subscriber account ("SaaS tenant").
 */
export enum DestinationOrigin {
  Subscriber = 'subscriber',
  Provider = 'provider'
}

/**
 * Declaration of a destination to be retrieved from an environment variable or from the destination service on SAP Cloud Platform.
 *
 * Use an object of this interface to specify which destination shall be used when executing a request.
 * The destination will be retrieved via its [[DestinationNameAndJwt.destinationName]] according to the following algorithm:
 * 1. If a destination of this [[DestinationNameAndJwt.destinationName]] is defined in the environment variable `destinations` (if available), it will be converted into a [[Destination]] and used for the request.
 * 2. Otherwise, the destination service on SAP Cloud Platform is queried for a destination with the given [[DestinationNameAndJwt.destinationName]], using the access token provided as value of property [[jwt]].
 */
export interface DestinationNameAndJwt {
  /**
   * Name of the destination to retrieve, mandatory.
   */
  destinationName: string;

  /**
   * An access token for the XSUAA service on SAP Cloud Platform, provided as a JSON Web Token, only mandatory when destination shall be retrieved from destination service on SAP Cloud Platform.
   */
  jwt?: string;
}

/**
 * @deprecated Since version 1.16.0. Use [[CachingOptions]] instead.
 */
export interface DestinationCachingOptions {
  /**
   * A boolean value that indicates whether to read destinations from cache.
   */
  useCache?: boolean;

  /**
   * The isolation strategy used for caching destinations. For the available options, see [[IsolationStrategy]].
   * By default, IsolationStrategy.Tenant is set.
   */
  isolationStrategy?: IsolationStrategy;
}

/**
 * Options to use while fetching destinations. Encompasses both [[DestinationCachingOptions]] and [[ResilienceOptions]] interfaces.
 */
export type DestinationRetrievalOptions = (
  | DestinationCachingOptions
  | CachingOptions
) &
  ResilienceOptions;

export function isDestinationNameAndJwt(
  destination: any
): destination is DestinationNameAndJwt {
  return destination.destinationName !== undefined;
}

export function isDestination(destination: any): destination is Destination {
  return destination.url !== undefined;
}

export type AuthenticationType =
  | 'PrincipalPropagation'
  | 'NoAuthentication'
  | 'BasicAuthentication'
  | 'OAuth2SAMLBearerAssertion'
  | 'OAuth2ClientCredentials'
  | 'OAuth2UserTokenExchange'
  | 'ClientCertificateAuthentication'
  | 'OAuth2JWTBearer'
  | 'OAuth2Password';

/**
 * The destinations endpoint distinguished between destinations maintained on service level (instance) and account level (subaccount).
 * This enum is used as a switch in the [[fetchInstanceDestinations]], [[fetchSubaccountDestinations]]  and [[destinationServiceCache]]
 */
export enum DestinationType {
  Instance = 'instance',
  Subaccount = 'subaccount'
}
