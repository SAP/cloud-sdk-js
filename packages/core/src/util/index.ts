export * from './jwt';
export * from './name-converter';
export * from './properties-util';
export * from './proxy-util';
export * from './remove-slashes';
