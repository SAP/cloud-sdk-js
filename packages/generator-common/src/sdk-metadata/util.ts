import { resolve } from 'path';
import { readFile } from 'fs-extra';

/**
 * Get the current SDK version from the package json.
 * @returns The SDK version.
 * @internal
 */
export async function getSdkVersion(): Promise<string> {
  return JSON.parse(
    await readFile(resolve(__dirname, '../../package.json'), 'utf8')
  ).version;
}
