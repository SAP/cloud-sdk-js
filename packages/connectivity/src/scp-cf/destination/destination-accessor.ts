import { createLogger, ErrorWithCause } from '@sap-cloud-sdk/util';
import { exchangeToken, isTokenExchangeEnabled } from '../identity-service';
import { parseSubdomain } from '../subdomain-replacer';
import {
  DestinationOrFetchOptions,
  sanitizeDestination,
  toDestinationNameUrl
} from './destination';
import { Destination } from './destination-service-types';
import { searchEnvVariablesForDestination } from './destination-from-env';
import {
  DestinationForServiceBindingOptions,
  searchServiceBindingForDestination
} from './destination-from-vcap';
import {
  getDestinationFromDestinationService,
  DestinationFromServiceRetriever,
  getDestinationServiceCredentials
} from './destination-from-service';
import {
  DestinationFetchOptions,
  isDestinationFetchOptions,
  AllDestinationOptions,
  DestinationWithoutToken
} from './destination-accessor-types';
import { searchRegisteredDestination } from './destination-from-registration';
import {
  fetchInstanceDestinations,
  fetchSubaccountDestinations
} from './destination-service';

const logger = createLogger({
  package: 'connectivity',
  messageContext: 'destination-accessor'
});

/**
 * Returns the parameter if it is a destination, calls {@link getDestination} otherwise (which will try to fetch the destination
 * from the Cloud Foundry destination service).
 *
 * Fetching a destination requires:
 * - A binding to exactly one XSUAA service instance with service plan "application".
 * - A binding to a destination service instance.
 *
 * If either of the prerequisites is not met or one of the services returns an error, this function will either throw an error or return a promise that rejects.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @returns A promise resolving to the requested destination on success.
 */
export async function useOrFetchDestination(
  destination: DestinationOrFetchOptions
): Promise<Destination | null> {
  return isDestinationFetchOptions(destination)
    ? getDestination(destination)
    : sanitizeDestination(destination);
}

/**
 * Resolve a destination by the following steps:
 * 1. Call [[useOrFetchDestination]]
 * 2. Throw an error, when the resulting destination from the previous step is falsy
 * 3. Return the checked destination.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @returns A promise resolving to the requested destination on success.
 * @internal
 */
export async function resolveDestination(
  destination: DestinationOrFetchOptions
): Promise<Destination> {
  const resolvedDestination = await useOrFetchDestination(destination).catch(
    error => {
      throw new ErrorWithCause('Failed to load destination.', error);
    }
  );
  if (!resolvedDestination) {
    throw Error(
      `Failed to resolve the destination '${toDestinationNameUrl(
        destination
      )}'.`
    );
  }
  return resolvedDestination;
}

/**
 * Builds a destination from one of three sources (in the given order):
 * - from the environment variable "destinations".
 * - from service bindings.
 * - from the destination service.
 *
 * If you want to get a destination only from a specific source, use the corresponding function directly
 * (`getDestinationFromEnvByName`, `destinationForServiceBinding`, `getDestinationFromDestinationService`).
 * @param options - The options to retrieve the destination.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestination(
  options: DestinationFetchOptions & DestinationForServiceBindingOptions
): Promise<Destination | null> {
  const destination =
    searchEnvVariablesForDestination(options) ||
    (await searchRegisteredDestination(options)) ||
    (await searchServiceBindingForDestination(options)) ||
    (await getDestinationFromDestinationService(options));
  return destination;
}

/**
 * Creates comprehensive log messages from a destinations array and their origin.
 * @param origin - Origin of the destination.
 * @param destinations - Array of destinations.
 * @returns Logs of the retrival of destinations.
 */
function createDestinationFetchLogs(
  origin: string,
  destinations: Destination[]
): string {
  return destinations.reduce(
    (prevLogMessages, currentDestination) =>
      prevLogMessages +
      `Retrieving ${origin} destination: ${currentDestination.name}.\n`,
    ''
  );
}

/**
 * Fetches all destinations from the destination service which match the token.
 * With a subscriber token it fetches all subscriber destinations, otherwise all provider destinations.
 * @param options - The {@link AllDestinationOptions | options} to fetch all destinations.
 * @returns A promise of an array of all destinations without authTokens from the destination service, on success.
 */
export async function getAllDestinationsFromDestinationService(
  options: AllDestinationOptions = {}
): Promise<DestinationWithoutToken[]> {
  logger.debug(
    'Attempting to retrieve all destinations from destination service.'
  );
  if (isTokenExchangeEnabled(options)) {
    options.jwt = await exchangeToken(options);
  }

  const token =
    (await DestinationFromServiceRetriever.getSubscriberToken(options))
      ?.serviceJwt ||
    (await DestinationFromServiceRetriever.getProviderServiceToken(options));

  const destinationServiceUri = getDestinationServiceCredentials().uri;
  const accountName = parseSubdomain(token.decoded.iss!);
  logger.debug(
    `Retrieving all destinations for account: "${accountName}" from destination service.`
  );

  const [instance, subaccount] = await Promise.all([
    fetchInstanceDestinations(destinationServiceUri, token.encoded, options),
    fetchSubaccountDestinations(destinationServiceUri, token.encoded, options)
  ]);

  const loggerMessage =
    createDestinationFetchLogs('instance', instance) +
    createDestinationFetchLogs('subaccount', subaccount);

  logger.debug(loggerMessage);

  const allDestinations = instance.concat(subaccount);

  if (allDestinations && allDestinations.length !== 0) {
    logger.debug(
      `Successfully retrieved all destinations for account: "${accountName}" from destination service.`
    );
  } else {
    logger.debug("Didn't receive any destinations from destination service.");
    return [];
  }

  const allDestinationsWithoutToken: DestinationWithoutToken[] =
    allDestinations.map(destination => {
      delete destination.authTokens;
      return destination;
    });

  return allDestinationsWithoutToken;
}
