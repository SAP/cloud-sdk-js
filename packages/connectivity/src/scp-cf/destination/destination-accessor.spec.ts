import { resolveDestinationWithType } from "./destination-accessor";
import { Destination } from "./destination-service-types";
import { pick } from "@sap-cloud-sdk/util";

describe('destination accessor', () => {
  const destination: Destination = {
    url: 'https://my.system.com',
    name: 'my-destination',
    proxyType: 'Internet',
    username: 'USER_NAME',
    password: 'password',
    authentication: 'BasicAuthentication',
    type: 'HTTP'
  };

  it('resolves destination when the destination type matches', async () => {
    const resolved = await resolveDestinationWithType(destination, 'HTTP');
    const keys = ['url', 'authentication', 'username', 'password'];
    expect(pick(keys, resolved)).toStrictEqual(pick(keys, destination));
  })

  it('throws an error, when destination type does not match', async () => {
    await expect(resolveDestinationWithType(destination, 'MAIL')).rejects.toThrowErrorMatchingSnapshot();
  })
})
