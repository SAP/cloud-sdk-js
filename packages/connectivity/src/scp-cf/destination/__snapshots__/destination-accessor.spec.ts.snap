// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`destination accessor throws an error, when destination type does not match 1`] = `"The type of the destination 'name: my-destination, url: https://my.system.com' has to be 'MAIL', but is 'HTTP'."`;
